### PR TITLE
Resolve: "ADD: FeeRates from thorchain"

### LINKED ISSUE #299: ADD: FeeRates from thorchain
# Thirdparty services are delayed

Instead of using third-parties, default to use thorchain gas oracle instead, which is always up to date

https://thornode.thorchain.info/thorchain/inbound_addresses

Gas rates you can get for each chain, if nothing there, then use the other service providers.

fastest: 1.5x 
fast: 1x
average: 0.67x

For PoW chains, the gas rate is a feeRate. 
For BNB, it is fixed (and not really relevant) but it is the total fee to pay

### PR DESCRIPTION
**Pull request by mikewyszinski**
Fix #299

### CODE REVIEW

**Comment by mikewyszinski** [comment]
refactored BCH, BTC and BNB to use the new base class... workign ETH and LTC next

**Comment by mikewyszinski** [comment]
refactored BCH, BTC and BNB to use the new base class... workign ETH and LTC next

**Review by veado — COMMENTED** [review]
@mikewyszinski Good stuff! Just some suggestions ...

BTW: Please add a more descriptive title to your PRs and link issues in its description properly. It saves some time to understand what's going on here ;)

File: packages/xchain-binance/src/client.ts [hunk 1]
```diff
@@ -448,28 +401,34 @@ class Client implements BinanceClient, XChainClient {
...
-      } as Fees
+      const rates = await this.getFeeRatesFromThorchain()
+      singleTxFee = baseAmount(rates.fast)
     } catch (error) {
-      return Promise.reject(error)
+      console.log(error)
+      console.warn(`Error pulling rates from thorchain, will try alternate`)
+    }
+    if (!singleTxFee) {
+      const transferFee = await this.getTransferFee()
```
[the remark this answers is not in the document]
Comment (veado) [reply packages/xchain-binance/src/client.ts:414]: All my change request to add `try/catch` blocks has been removed (based on @reidrankin feedback https://discord.com/channels/838986635756044328/842252939149967382/859092020097056829)

File: packages/xchain-binance/src/client.ts [hunk 2]
```diff
@@ -220,16 +177,16 @@ class Client implements BinanceClient, XChainClient {
    *
    * @throws {Error} Thrown if phrase has not been set before. A phrase is needed to create a wallet and to derive an address from it.
    */
-  getAddress = (index = 0): string =>
-    crypto.getAddressFromPrivateKey(this.getPrivateKey(index), getPrefix(this.network))
-
+  public getAddress(index = 0): string {
```
[the remark this answers is not in the document]
Comment (mikewyszinski) [reply packages/xchain-binance/src/client.ts:180]: agree. i added them trying to figure out some comile errors re: the bae class.  i removed them all

File: packages/xchain-client/src/BaseXChainClient.ts [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+  abstract getAddress(walletIndex: number): string
+  abstract getExplorerUrl(): string
+  abstract getExplorerAddressUrl(address: string): string
+  abstract getExplorerTxUrl(txID: string): string
+  abstract validateAddress(address: string): boolean
+  abstract getBalance(address: string, assets?: Asset[]): Promise<Balances>
+  abstract getTransactions(params?: TxHistoryParams): Promise<TxsPage>
+  abstract getTransactionData(txId: string, assetAddress?: string): Promise<Tx>
+  abstract transfer(params: TxParams): Promise<string>
+  // abstract purgeClient(): void
```
[the remark this answers is not in the document]
Comment (mikewyszinski) [reply packages/xchain-client/src/BaseXChainClient.ts:136]: nope. removed

File: packages/xchain-binance/src/client.ts [hunk 4]
```diff
@@ -448,28 +401,34 @@ class Client implements BinanceClient, XChainClient {
...
-      const singleTxFee = baseAmount(transferFee.fixed_fee_params.fee)
-
-      return {
-        type: 'base',
-        fast: singleTxFee,
-        fastest: singleTxFee,
-        average: singleTxFee,
-      } as Fees
+      const rates = await this.getFeeRatesFromThorchain()
+      singleTxFee = baseAmount(rates.fast)
```
[the remark this answers is not in the document]
Comment (mikewyszinski) [reply packages/xchain-binance/src/client.ts:408]: i think letting it default to thorchain's 8 is correct in this case, let me know if you still disagree

**Comment by mikewyszinski** [comment]
@veado other that that last item above, ready to merge and release

File: packages/xchain-ethereum/__tests__/client.testnet.test.ts [hunk 5]
```diff
@@ -353,14 +361,32 @@ describe('Client Test', () => {
...
+        gas_rate: '49',
+      },
+    ])
+
+    // mock_gastracker_api(etherscanUrl, 'gasoracle', {
+    //   LastBlock: '11745402',
+    //   SafeGasPrice: '51',
+    //   ProposeGasPrice: '59',
+    //   FastGasPrice: '76',
+    // })
```
Comment (veado) [inline packages/xchain-ethereum/__tests__/client.testnet.test.ts]: [nitpick] Still needed ^?

File: packages/xchain-ethereum/__tests__/client.testnet.test.ts [hunk 6]
```diff
@@ -563,14 +591,31 @@ describe('Client Test', () => {
...
+        halted: false,
+        gas_rate: '49',
+      },
+    ])
+    // mock_gastracker_api(etherscanUrl, 'gasoracle', {
+    //   LastBlock: '11745402',
+    //   SafeGasPrice: '51',
+    //   ProposeGasPrice: '59',
+    //   FastGasPrice: '76',
+    // })
```
Comment (veado) [inline packages/xchain-ethereum/__tests__/client.testnet.test.ts]: ^ Still needed?

File: packages/xchain-binance/src/client.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (veado) [reply packages/xchain-binance/src/client.ts:408]: That's ok for now (was just thinking if anybody wants to change default decimal in `BaseAmount` / `AssetAmount` for any reason in the future...)

**Review by veado — APPROVED** [review]
LGTM

File: packages/xchain-ethereum/__tests__/client.testnet.test.ts (hunk 5 above)
Comment (mikewyszinski) [reply packages/xchain-ethereum/__tests__/client.testnet.test.ts]: reseoved

File: packages/xchain-ethereum/__tests__/client.testnet.test.ts (hunk 6 above)
Comment (mikewyszinski) [reply packages/xchain-ethereum/__tests__/client.testnet.test.ts]: resolved

**Pull request merged**